File: app/test/operately/rich_content/resource_links_test.exs
defmodule Operately.RichContent.ResourceLinksTest do
  use Operately.DataCase, async: true

  alias Operately.RichContent.ResourceLinks

  @origin "https://app.operately.com"
  @company "acme-0abc"

  test "URL labels ignore fragments and trailing slashes while preserving custom labels" do
    href = "#{@origin}/#{@company}/projects/website-xyz?tab=overview#comment"
    assert ResourceLinks.url_label?(href, href)
    assert ResourceLinks.url_label?(String.replace(href, "#comment", "#other"), href)
    assert ResourceLinks.url_label?("#{@origin}/#{@company}/projects/website-xyz/", href)
    assert ResourceLinks.url_label?("/#{@company}/projects/website-xyz", "/#{@company}/projects/website-xyz#comment")
    refute ResourceLinks.url_label?("Website", href)
    refute ResourceLinks.url_label?(nil, href)
  end

  test "rejects non-web schemes and accepts same-origin protocol-relative links" do
    assert :error = ResourceLinks.parse("javascript:/#{@company}/projects/website-xyz", @origin)
    assert {:ok, %{type: :project}} = ResourceLinks.parse("//app.operately.com/#{@company}/projects/website-xyz", @origin)
  end

  describe "parse/2" do
    test "extracts company, type, and id from recognized resource routes" do
      assert {:ok, %{company_id: @company, type: :project, id: "website-xyz"}} =
               ResourceLinks.parse("#{@origin}/#{@company}/projects/website-xyz", @origin)

      assert {:ok, %{type: :task, id: "ship-it"}} = ResourceLinks.parse("#{@origin}/#{@company}/tasks/ship-it", @origin)
      assert {:ok, %{type: :goal, id: "nps"}} = ResourceLinks.parse("#{@origin}/#{@company}/goals/nps", @origin)
      assert {:ok, %{type: :milestone, id: "beta"}} = ResourceLinks.parse("#{@origin}/#{@company}/milestones/beta", @origin)
      assert {:ok, %{type: :discussion, id: "launch"}} = ResourceLinks.parse("#{@origin}/#{@company}/discussions/launch", @origin)
      assert {:ok, %{type: :document, id: "abc"}} = ResourceLinks.parse("#{@origin}/#{@company}/documents/abc", @origin)
      assert {:ok, %{type: :link, id: "abc"}} = ResourceLinks.parse("#{@origin}/#{@company}/links/abc", @origin)
      assert {:ok, %{type: :file, id: "abc"}} = ResourceLinks.parse("#{@origin}/#{@company}/files/abc", @origin)
      assert {:ok, %{type: :folder, id: "abc"}} = ResourceLinks.parse("#{@origin}/#{@company}/folders/abc", @origin)
      assert {:ok, %{type: :person, id: "jane"}} = ResourceLinks.parse("#{@origin}/#{@company}/people/jane", @origin)
      assert {:ok, %{type: :space, id: "product"}} = ResourceLinks.parse("#{@origin}/#{@company}/spaces/product", @origin)
    end

    test "accepts relative paths on the current instance" do
      assert {:ok, %{company_id: @company, type: :project, id: "website-xyz"}} =
               ResourceLinks.parse("/#{@company}/projects/website-xyz", @origin)
    end

    test "extracts space kanban task ids" do
      url = "#{@origin}/#{@company}/spaces/product-space/kanban?taskId=ship-it"

      assert {:ok, %{type: :task, id: "ship-it"}} = ResourceLinks.parse(url, @origin)
    end

    test "strips fragments while keeping the resource id" do
      url = "#{@origin}/#{@company}/projects/website-xyz?tab=tasks#comment-1"

      assert {:ok, %{type: :project, id: "website-xyz"}} = ResourceLinks.parse(url, @origin)
    end

    test "rejects other origins, nested paths, and unknown resource types" do
      assert :error = ResourceLinks.parse("https://other.example/#{@company}/projects/website-xyz", @origin)
      assert :error = ResourceLinks.parse("#{@origin}/#{@company}/projects/website-xyz/pause", @origin)
      assert :error = ResourceLinks.parse("#{@origin}/#{@company}/work-map", @origin)
      assert :error = ResourceLinks.parse("https://example.com", @origin)
    end
  end
end

File: turboui/src/RichContent/resourceLinks.test.ts
import { applyResourceLinkTitles, collectResourceLinkRefs, isUrlLabel, parseResourceLinkUrl } from "./resourceLinks";

const origin = "https://app.operately.com";
const companyId = "acme-0abc";
const options = { origin, companyId };

function linkDoc(text: string, href: string) {
  return {
    type: "doc",
    content: [
      {
        type: "paragraph",
        content: [
          {
            type: "text",
            text,
            marks: [{ type: "link", attrs: { href, target: "_blank" } }],
          },
        ],
      },
    ],
  };
}

describe("resourceLinks", () => {
  describe("parseResourceLinkUrl", () => {
    it("extracts company, type, and id from recognized routes", () => {
      expect(parseResourceLinkUrl(`${origin}/${companyId}/projects/website-xyz`, options)).toEqual({
        companyId,
        type: "project",
        id: "website-xyz",
      });
    });

    it("accepts relative paths and space kanban task urls", () => {
      expect(parseResourceLinkUrl(`/${companyId}/goals/nps`, options)).toEqual({
        companyId,
        type: "goal",
        id: "nps",
      });

      expect(parseResourceLinkUrl(`${origin}/${companyId}/spaces/product/kanban?taskId=ship-it`, options)).toEqual({
        companyId,
        type: "task",
        id: "ship-it",
      });
    });

    it("rejects other origins, companies, and nested paths", () => {
      expect(parseResourceLinkUrl("https://other.example/acme-0abc/projects/website-xyz", options)).toBeNull();
      expect(parseResourceLinkUrl(`${origin}/other-0def/projects/website-xyz`, options)).toBeNull();
      expect(parseResourceLinkUrl(`${origin}/${companyId}/projects/website-xyz/pause`, options)).toBeNull();
    });
  });

  describe("collectResourceLinkRefs", () => {
    it("collects unique url-labeled resource links and ignores custom labels", () => {
      const href = `${origin}/${companyId}/projects/website-xyz`;
      const content = {
        type: "doc",
        content: [
          {
            type: "paragraph",
            content: [
              {
                type: "text",
                text: href,
                marks: [{ type: "link", attrs: { href } }],
              },
              {
                type: "text",
                text: href,
                marks: [{ type: "link", attrs: { href: `${href}?tab=overview` } }],
              },
              {
                type: "text",
                text: "Custom label",
                marks: [{ type: "link", attrs: { href } }],
              },
              {
                type: "text",
                text: "https://example.com",
                marks: [{ type: "link", attrs: { href: "https://example.com" } }],
              },
            ],
          },
        ],
      };

      expect(collectResourceLinkRefs(content, options)).toEqual([{ type: "project", id: "website-xyz" }]);
    });
  });

  describe("applyResourceLinkTitles", () => {
    it("replaces url labels with titles and leaves custom labels and destinations unchanged", () => {
      const href = `${origin}/${companyId}/projects/website-xyz?tab=tasks#notes`;
      const content = linkDoc(href, href);
      const custom = linkDoc("Custom label", href);

      const titled = applyResourceLinkTitles(content, [{ type: "project", id: "website-xyz", title: "Website" }], {
        origin,
      }) as { content: { content: { text: string; marks: { attrs: { href: string } }[] }[] }[] };

      expect(titled.content[0]?.content[0]?.text).toBe("Website");
      expect(titled.content[0]?.content[0]?.marks[0]?.attrs.href).toBe(href);

      const unchanged = applyResourceLinkTitles(custom, [{ type: "project", id: "website-xyz", title: "Website" }], {
        origin,
      }) as { content: { content: { text: string }[] }[] };

      expect(unchanged.content[0]?.content[0]?.text).toBe("Custom label");
      expect(content.content[0]?.content[0]?.text).toBe(href);
    });

    it("keeps the original url when no title is available", () => {
      const href = `${origin}/${companyId}/projects/website-xyz`;
      const content = linkDoc(href, href);

      expect(applyResourceLinkTitles(content, [], { origin })).toBe(content);
    });
  });

  describe("isUrlLabel", () => {
    it("treats the href and href without hash as url labels", () => {
      const href = `${origin}/${companyId}/projects/website-xyz?tab=tasks#notes`;

      expect(isUrlLabel(href, href)).toBe(true);
      expect(isUrlLabel(`${origin}/${companyId}/projects/website-xyz?tab=tasks`, href)).toBe(true);
      expect(isUrlLabel("Website", href)).toBe(false);
    });
  });
});
